### PR TITLE
Added fast path in the case of no hooks

### PR DESCRIPTION
**Pull request by Chillee**
See the discussion here: https://github.com/pytorch/pytorch/pull/50431

~~Not completely done yet - need to figure out the backwards compatibility stuff as well as `RemovableHandle`.~~

~~Also, this concretely breaks Torchscript (which tries to script the properties), and more generally, probably requires modifying Torchscript hook support: https://github.com/pytorch/pytorch/issues/34329~~

Just kidding, I think all problems are solved :)

Another thing I could do in this PR is to simply replace all the `len(x) > 0` checks with the faster checks. That's about 1.5-2k more Python instructions and .4 - .5 microseconds slower.

### CODE REVIEW

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -920,8 +946,6 @@ def _call_impl(self, *input, **kwargs):
     def __setstate__(self, state):
         self.__dict__.update(state)
         # Support loading old checkpoints that don't have the following attrs:
-        if '_forward_pre_hooks' not in self.__dict__:
```
Comment (Chillee) [inline torch/nn/modules/module.py:923]: Not sure what the right thing to do here is.

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -864,6 +864,10 @@ def _slow_forward(self, *input, **kwargs):
         return result
 
     def _call_impl(self, *input, **kwargs):
+        forward_call = (self._slow_forward if torch._C._get_tracing_state() else self.forward)
```
Comment (albanD) [inline torch/nn/modules/module.py:867]: Could you add a comment here to explain what this function is doing?
In particular, write explicitly that if no hook is there, just skip the whole function. Or maybe actually rename this into `_call_with_hooks()` and make the parent function skip it if there are no hooks at all in this Module?

**Review by albanD — COMMENTED** [review]
The change sounds good to me.
Just a small comment about potential refactoring.

File: torch/nn/modules/module.py [hunk 3]
```diff
@@ -864,41 +864,45 @@ def _slow_forward(self, *input, **kwargs):
...
+                    self._forward_pre_hooks.values()):
+                result = hook(self, input)
+                if result is not None:
+                    if not isinstance(result, tuple):
+                        result = (result,)
+                    input = result
 
         bw_hook = None
-        if len(full_backward_hooks) > 0:
+        if full_backward_hooks:
```
Comment (albanD) [inline torch/nn/modules/module.py:888]: This is not defined in the case where these is no backward hooks right? You want to initialize it to None I think.

File: torch/nn/modules/module.py [hunk 4]
```diff
@@ -864,41 +864,45 @@ def _slow_forward(self, *input, **kwargs):
...
+        result = forward_call(*input, **kwargs)
+        if _global_forward_hooks or self._forward_hooks:
+            for hook in itertools.chain(
+                    _global_forward_hooks.values(),
+                    self._forward_hooks.values()):
+                hook_result = hook(self, input, result)
+                if hook_result is not None:
+                    result = hook_result
 
         if bw_hook:
```
Comment (albanD) [inline torch/nn/modules/module.py:901]: nit: you can re-use full_backward_hooks here to avoid setting `bw_hook=None` all the time.

**Review by albanD — COMMENTED** [review]
LGTM
Just some variable initialization issue and this should be good to go I think!

Thanks for the update for the "slow path" as well.

File: torch/nn/modules/module.py (hunk 3 above)
Comment (Chillee) [reply torch/nn/modules/module.py:888]: No, that's the main thing I discovered regarding `if <list>` or `if dict`.

```
print(bool([]), bool([1])
>>> False True
```

Comment (albanD) [reply torch/nn/modules/module.py:888]: Ho we already pre-define it as an empty list. My bad!

**Review by albanD — APPROVED** [review]
I read the existing code too fast. The variable is already properly initialized. LGTM!

**Pull request closed without merging**